provider/cluster/state.go

```diff
@@ -0,0 +1,188 @@
+	}
+	go runLoop(ds, stateDeploying, log)
+	return ds
+}
+
+func newTestDeployState(ctx context.Context, s *service, lease mtypes.LeaseID, mgroup *manifest.Group) (*deployState, stateFunc) {
+	log := s.log.With("cmp", "deployment-state", "lease", lease, "manifest-group", mgroup.Name)
+
+	ds := &deployState{
+		ctx:     ctx,
+		bus:     s.bus, // unused
+		client:  s.client,
+		session: s.session, // unused
+
+		state:  stateDeploying, // initial state, deploy the manifest
+		lease:  lease,
+		mgroup: mgroup,
+
+		updatech:   make(chan *manifest.Group),
+		teardownch: make(chan struct{}),
+		errch:      nil,
+
+		log: log.With("cmp", "deploystate"),
+	}
+	var sf stateFunc
+	initState := stateDeploying
+	go func(initState stateFunc) {
+		state := initState
+		for state != nil {
+			sf = state(ds)
+			log.With("state", s).Debug("state transitioned")
+			state = sf
+		}
+	}(initState) // runLoop(ds, stateDeploying, log)
+	return ds, sf
+}
+
+func (ds *deployState) run() {
+	// For clean separation declare and use state as a local variable. eg: `state := ds.startState`
+	for ds.state != nil {
+		ds.state = ds.state(ds)
+	}
+}
+
+func (ds *deployState) update(mgroup *manifest.Group) error {
+	ds.mux.Lock()
+	defer ds.mux.Unlock()
+
+	ds.errch = make(chan error, 1)
+	ds.updatech <- mgroup
+	return <-ds.errch
+}
+
+func (ds *deployState) teardown() error {
+	ds.mux.Lock()
+	defer ds.mux.Unlock()
+
+	ds.errch = make(chan error, 1)
+	ds.teardownch <- struct{}{}
+	return <-ds.errch
+
+}
+
+// runLoop provides state transition loop until there are no more states.
+// logs transitions between states, and initializes the provided state with
+// the deployState struct.
+func runLoop(ds *deployState, initState stateFunc, log log.Logger) {
+	state := initState
+	for state != nil {
+		s := state(ds)
+		log.With("state", s).Debug("state transitioned")
+		state = s
+	}
+}
+
+// deployStateStart initializes the deployment and waits till it is fully operational
+func stateDeploying(ds *deployState) stateFunc {
+	if ds.monitor != nil {
+		ds.monitor.shutdown() //
+		ds.monitor = nil
+	}
+	// TODO: Pass ds context to Deploy()
+	err := ds.client.Deploy(ds.ctx, ds.lease, ds.mgroup)
+	if err != nil {
+		ds.log.Error("deploying error", err)
+		if ds.errch != nil {
+			ds.errch <- err
+		}
+		return stateErrored
+	}
+
+	ds.monitor = newDeploymentMonitor(ds.ctx, ds.bus, ds.session, ds.client, ds.lease, ds.mgroup, ds.log)
+	return stateNominal
+}
+
+// stateNominal is entered when deployment is active, and there is nothing to do
+// but wait for signals to perform actions.
+func stateNominal(ds *deployState) stateFunc {
+	select {
+	case mg := <-ds.updatech:
+		ds.mgroup = mg
+		return stateDeploying
+	case <-ds.ctx.Done():
+		return stateTearingdown
+	case <-ds.teardownch:
+		return stateTearingdown
+	}
+}
+
+// stateErrored engages when an uncrecoverable point has been reached.
+func stateErrored(ds *deployState) stateFunc {
+	ds.errch = nil // reset the error channel
```

Comment:
race condition.  consider this series of events:

1. `update()` is called, the update fails and state gets set to `stateErrored`.
2. as `stateErrored` begins executing, someone calls `update()` or `teardown()`.